ios/AirGoogleMaps/AIRGoogleMap.m

```diff
@@ -0,0 +1,222 @@
+//
+//  AIRGoogleMap.m
+//  AirMaps
+//
+//  Created by Gil Birman on 9/1/16.
+//
+
+#import "AIRGoogleMap.h"
+#import "AIRGoogleMapMarker.h"
+#import <GoogleMaps/GoogleMaps.h>
+#import <MapKit/MapKit.h>
+#import "RCTConvert+MapKit.h"
+#import "UIView+React.h"
+
+id cameraPositionAsJSON(GMSCameraPosition *position) {
+  // todo: convert zoom to delta lat/lng
+  return @{
+           @"latitude": [NSNumber numberWithDouble:position.target.latitude],
+           @"longitude": [NSNumber numberWithDouble:position.target.longitude],
+           @"zoom": [NSNumber numberWithDouble:position.zoom],
+           };
+}
+
+@interface AIRGoogleMap ()
+
+- (id)eventFromCoordinate:(CLLocationCoordinate2D)coordinate;
+
+@end
+
+@implementation AIRGoogleMap
+{
+  NSMutableArray<UIView *> *_reactSubviews;
+  BOOL _initialRegionSet;
+}
+
+- (instancetype)init
+{
+  if ((self = [super init])) {
+    _reactSubviews = [NSMutableArray new];
+    _markers = [NSMutableArray array];
+    _initialRegionSet = false;
+  }
+  return self;
+}
+- (id)eventFromCoordinate:(CLLocationCoordinate2D)coordinate {
+
+  CGPoint touchPoint = [self.projection pointForCoordinate:coordinate];
+
+  return @{
+           @"coordinate": @{
+               @"latitude": @(coordinate.latitude),
+               @"longitude": @(coordinate.longitude),
+               },
+           @"position": @{
+               @"x": @(touchPoint.x),
+               @"y": @(touchPoint.y),
+               },
+           };
+}
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wobjc-missing-super-calls"
+- (void)insertReactSubview:(id<RCTComponent>)subview atIndex:(NSInteger)atIndex {
+  // Our desired API is to pass up markers/overlays as children to the mapview component.
+  // This is where we intercept them and do the appropriate underlying mapview action.
+  if ([subview isKindOfClass:[AIRGoogleMapMarker class]]) {
+    AIRGoogleMapMarker *marker = (AIRGoogleMapMarker*)subview;
+    marker.realMarker.map = self;
+    [self.markers addObject:marker];
+  }
+  [_reactSubviews insertObject:(UIView *)subview atIndex:(NSUInteger) atIndex];
+}
+#pragma clang diagnostic pop
+
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wobjc-missing-super-calls"
+- (void)removeReactSubview:(id<RCTComponent>)subview {
+  // similarly, when the children are being removed we have to do the appropriate
+  // underlying mapview action here.
+  if ([subview isKindOfClass:[AIRGoogleMapMarker class]]) {
+    AIRGoogleMapMarker *marker = (AIRGoogleMapMarker*)subview;
+    marker.realMarker.map = nil;
+    [self.markers removeObject:marker];
+  }
+  [_reactSubviews removeObject:(UIView *)subview];
```

Comment:
what happens if we call `[super removeReactSubview:subview]` are we intentionally *not* calling it here?